### ISSUE TITLE
Adding an easy way to display error messages

### ISSUE THREAD

**Issue by kaw2k** [willowtreeapps/react-formable#11]
# Showing errors within forms

Currently, this is how we display errors:

```js
const LoginForm = React.createClass({
    getInitialState() {
        return { errors: [] };
    },
    
    onSubmit(form) {
        const errors = this.refs.form.showFieldErrors();
    },

    render() {
        const errors = this.state.errors.map((e, i) => <li key={i}>{e}</li>)
        
        return <Form ref="form" onSubmit={this.onSubmit}>
            <Input name="username" type="text" />
            <Input name="password" type="text" />
            <button>Submit</button>
            <ul className="errors">{errors}</ul>
        </Form>;
    }
});
```
We just slap a `ref` on the `Form` and call its `showFieldErrors` method. This returns an array of errors which we then save to our `state`. This forces a re-render which displays a list of errors returned by `showFieldErrors`. 

This isn't _toooooo_ bad, however every form realistically will want to display errors. We can't dictate how those errors will be shown. However, we need to make it easier to show errors. We don't want every form implementing state just to display errors when the `Form` already keeps track of that information for us.

## Possible idea: `Errors` component
Rendering errors to the screen is something that is screaming to be a component. For example:

```js
const Errors = React.createClass({
    render() {
        const errors = this.props.errors || [];
        
        return <ul className="errors">
            {errors.map((e, i) => <li key={i}>{e}</li>)}
        </ul>
    }
});
```
We just took out the rendering of the errors and threw it into a component. We can then integrate this into forms in two ways:

#### 1) Reserve a `name`
We can change how form's pass down props to accommodate for reserved names. Let's say we choose `fieldErrors`.

```js
const LoginForm = React.createClass({
    render() {    
        return <Form ref="form" onSubmit={this.onSubmit}>
            <Input name="username" type="text" />
            <Input name="password" type="text" />
            <button>Submit</button>
            <Errors name="fieldErrors" />
        </Form>;
    }
});
```
Now, whenever our form encounters a component with a name of `fieldErrors` it will pass down as props `errors`. This name field could be customized (or not). 

In my mind there are a few issues here. My biggest concern is that there is too much mental overhead for the user. They need to add a special component AND a special name. Which brings us to our second solution...

#### 2) Sniff the component type
We can get around using a name by sniffing the component type. Now, it is just the following:

```js
const LoginForm = React.createClass({
    render() {    
        return <Form ref="form" onSubmit={this.onSubmit}>
            <Input name="username" type="text" />
            <Input name="password" type="text" />
            <button>Submit</button>
            <Errors />
        </Form>;
    }
});
```
There are other issues, now our form is wedded to this exact component (bad). No wrapping or extensibility possible. 

## Possible idea: Add a new prop for the `Form` to monitor
Similar to how we have `name`, we could have the `Form` look for another property. Hypothetically, lets say the `Form` listens for `for` where its value is the name of the error field to listen for. The implementation could either pass down errors as props, or wrap the element and put the errors adjacent to it. 

```js
const LoginForm = React.createClass({
    render() {    
        return <Form ref="form" onSubmit={this.onSubmit}>
            <Input name="username" type="text" />
            <Input name="password" type="text" />
            <button>Submit</button>
            {/* Option 1: Pass down errors */}
            <Errors for="errors" />

            {/* Option 2: Wrap the element */}
            <div for="errors" />
            {/* Turns into
                <div>
                    <div for="errors" />
                    <div className="errors">...</div>
                </div>
            */}
        </Form>;
    }
});
```
I am not a fan of modifying the DOM, but throwing it out there. Overall, introducing another prop type adds complexity, but could open the door for new interesting functionality. I still am leaning with the first idea.

**Comment by kaw2k** [comment]
Thinking about this more (and talking with @tlegard), sniffing the component type is not as bad as it seems to be. Any component that wants to extend this component would just need to implement a displayName of `Errors` (or something on the prototype or statics) and it would soak up the errors. 

The more I think about this, the easier it is for the user the better. If the 90% is very easy, we win. I don't see many people extending this component anyhow.

**Comment by kaw2k** [comment]
A follow up question, how would we expect this to behave with nested data? For example, lets say we are using `Fieldset` or `Fieldlist`:

```js
<Form>
    <Input name="name" />
    <Fieldset name="temp">
        <Input name="fun" />
        <Errors />
    </Fieldset>
    <Errors />
</Form>
```

We have two `Errors` components here, one is nested in a `Fieldset`. Do we expect the `Errors` component nested in the `Fieldset` to only get errors regarding the `Fieldset`? This could be useful to pool grouped errors together. 

The alternative is all form errors get passed down always.

**Comment by kaw2k** [comment]
Thinking of this more, we can probably accommodate both. Imagine the following props for `<Errors />`

prop | value | description
----- | ------- | ---------------- 
renderError | `[function -> node]` | A optional function that returns an error to render. Defaults to `function(x) { return x; }`.
additionalErrors | `[array[string]]` | An additional list of errors to render. Useful for rendering business level logic errors. Defaults to `[]`.
scoped | `boolean` | A flag to decide what level errors to render. Errors are scoped to the owning component. Global errors are errors the entire form owns. Defaults to `false`.

With this in mind, let's revisit the above example: 

```js
<Form>
    <Input name="name" />
    <Fieldset name="temp">
        <Input name="fun" />
        <Errors scoped /> {/* 1 */}
    </Fieldset>
    <Errors scoped /> {/* 2 */}
    <Errors /> {/* 3 */}
</Form>
```

`Errors1` will receive errors for `fun`
`Errors2` will receive errors for `name` and `temp`
`Errors3` will receive errors for `name`, `temp`, and `fun`

**Comment by kaw2k** [comment]
Addressed in commit 23bab79. Still needs some love, `scoped` needs work. Should do for now.
